TrackletAlgorithm/ProjectionRouter.h

```diff
@@ -75,134 +70,138 @@ void ProjectionRouter(BXType bx,
-      // hourglass configuration
+        //////////////////////////
+        // hourglass configuration
    
-      // top 5 bits of phi
-      auto iphi5 = iphiproj>>(iphiproj.length()-5);
+        // top 5 bits of phi
+        auto iphi5 = iphiproj>>(iphiproj.length()-5);
     
-      // total number of VMs
-      auto nvm = LAYER!=0 ? nvmmelayers[LAYER-1]*nallstubslayers[LAYER-1] :
-        nvmmedisks[DISK-1]*nallstubsdisks[DISK-1];
-      //assert(nvm>0);
+        // total number of VMs
+        auto nvm = LAYER!=0 ? nvmmelayers[LAYER-1]*nallstubslayers[LAYER-1] :
+          nvmmedisks[DISK-1]*nallstubsdisks[DISK-1];
+        //assert(nvm>0);
     
-      // number of VMs per module
-      auto nbins = LAYER!=0 ? nvmmelayers[LAYER-1] : nvmmedisks[DISK-1];
+        // number of VMs per module
+        auto nbins = LAYER!=0 ? nvmmelayers[LAYER-1] : nvmmedisks[DISK-1];
 
-      // routing
-      ap_uint<3> iphi = (iphi5/(32/nvm))&(nbins-1);  // OPTIMIZE ME
+        // routing
+        ap_uint<3> iphi = (iphi5/(32/nvm))&(nbins-1);  // OPTIMIZE ME
     
-      ///////////////
-      // VMProjection
-      static_assert(not DISK, "PR: Layer only for now.");
+        ///////////////
+        // VMProjection
+        static_assert(not DISK, "PR: Layer only for now.");
     
-      // vmproj index
-      typename VMProjection<VMPTYPE>::VMPID index = i;
-      //assert(i < (1<<index.length()));
-
-      // vmproj z
-      // Separate the vm projections into zbins
-      // The central bin e.g. zbin=4+(zproj.value()>>(zproj.nbits()-3));
-      // (assume 8 bins; take top 3 bits of zproj and shift it to make it positive)
-      // But we need some range (particularly for L5L6 seed projecting to L1-L3):
-      // Lower bound
-      ap_uint<MEBinsBits+1> zbin1tmp = (1<<(MEBinsBits-1))+(((izproj>>(izproj.length()-MEBinsBits-2))-2)>>2);
-      // Upper bound
-      ap_uint<MEBinsBits+1> zbin2tmp = (1<<(MEBinsBits-1))+(((izproj>>(izproj.length()-MEBinsBits-2))+2)>>2);
-      if (zbin1tmp >= (1<<MEBinsBits)) zbin1tmp = 0; //note that zbin1 is unsigned
-      if (zbin2tmp >= (1<<MEBinsBits)) zbin2tmp = (1<<MEBinsBits)-1;
-
-      // One extra bit was used in the above calculation. Now take it away.
-      ap_uint<MEBinsBits> zbin1 = zbin1tmp;
-      ap_uint<MEBinsBits> zbin2 = zbin2tmp;
-      //assert(zbin1<=zbin2);
-      //assert(zbin2-zbin1<=1);
+        // vmproj index
+        typename VMProjection<VMPTYPE>::VMPID index = nallproj;
+        //assert(i < (1<<index.length()));
+
+        // vmproj z
+        // Separate the vm projections into zbins
+        // The central bin e.g. zbin=4+(zproj.value()>>(zproj.nbits()-3));
+        // (assume 8 bins; take top 3 bits of zproj and shift it to make it positive)
+        // But we need some range (particularly for L5L6 seed projecting to L1-L3):
+        // Lower bound
+        ap_uint<MEBinsBits+1> zbin1tmp = (1<<(MEBinsBits-1))+(((izproj>>(izproj.length()-MEBinsBits-2))-2)>>2);
```

Comment:
I also don't find the comment very helpful. To some extent, I think we've inherited some of the messiness from the emulation, c.f.:
https://github.com/cms-L1TK/cmssw/blob/L1TK-dev-11_2_0_pre6/L1Trigger/TrackFindingTracklet/src/LayerProjection.cc#L70-L73
But if @djcranshaw understands what these lines are doing, please update the comment.

And if there's a way to do this more intuitively with, e.g., the `ap_int::range()` method, we should probably switch to that. I can't tell if that's possible right now, because it's hard to grasp what the line is doing.

Reply:
@djcranshaw are you going to address this?

Reply:
Apologies, I have been a bit backed up but I will look at these now.

Reply:
Okay, so it took a bit of time to figure out what these lines were actually doing. I've added a few comments which will hopefully explain it better, and I've replaced the many 2's with named constants (some of the twos are not the same as the other twos, which is precisely the sort of situation we want to avoid by having named constants in the first place). Basically what the lines do is take the top 3 (MEBinsBits) + 2 (zbins_extrabits) of zproj, and adjusting the projection in z up and down by 2 (zbins_adjust), then truncating the last 2 (zbins_extrabits) of zproj to get the zbin. Then if the bins are the same, the ME has to only look in that bin, but if they're different it has to look in both for stubs that match the projection.

Reply:
Another thing, I tried to improve the readability of these lines by replacing the bitshifts with ranges, but it wasn't so straightforward because izproj is signed

Reply:
Okay I'm now finished with these two lines. With the latest commit, the logic in these lines is much easier to understand.